sklearn/cluster/_bicluster.py

```diff
@@ -30,8 +30,8 @@ def _scale_normalize(X):
     X = make_nonnegative(X)
     row_diag = np.asarray(1.0 / np.sqrt(X.sum(axis=1))).squeeze()
     col_diag = np.asarray(1.0 / np.sqrt(X.sum(axis=0))).squeeze()
-    row_diag = np.where(np.isnan(row_diag), 0, row_diag)
-    col_diag = np.where(np.isnan(col_diag), 0, col_diag)
+    row_diag = np.nan_to_num(row_diag)
+    col_diag = np.nan_to_num(col_diag)
```

Comment:
We probably need to check the reference paper for biclustering to see if they recommend a particular way to deal with 0-sum rows and columns in `X`. I see two possible alternatives:

- what you do here, which would result in multiplication by huge values in the following lines;
- alternatively, use a neutral scale value of 1.0 for entries that match 0-sum rows and columns of `X`.

In any case, we probably need a dedicated non-regression test that checks for this edge-case explicitly in a biclustering context, rather than just relying on the generic estimator checks such as `check_estimator_sparse_array` and `check_estimators_dtypes` that are blind to the mathematical meaning of what is a valid biclustering result for 0 valued rows or columns.

Reply:
Unfortunately there is no particular way recommended to deal with 0-sum rows and columns in `X` in the original paper. There could be some subsequent work done by the same or a research group with a use case that fits this particular edge case more - should I look for any clues? Or should we continue by implementing any of the alternatives and write more specialised tests?

Reply:
Both examples produce no visible difference (only slight variations in the shuffled dataset image).
Code for the approximation: 
```
row_diag = np.nan_to_num(row_diag)
col_diag = np.nan_to_num(col_diag)
```

Code for replacing `inf` as the result of 0-sum-rows with `1.0`:
```
#Keeping these two lines as they were
row_diag = np.where(np.isnan(row_diag), 0, row_diag)
col_diag = np.where(np.isnan(col_diag), 0, col_diag)
#Adding the replacements for inf
row_diag = np.where(np.isinf(row_diag), 1.0, row_diag)
col_diag = np.where(np.isinf(col_diag), 1.0, col_diag)
```

<img width="663" alt="Screenshot 2024-10-08 at 11 36 41" src="https://github.com/user-attachments/assets/652fb17c-8905-42b9-a412-c248874a8401">

Reply:
The visual results show not discernible visual differences for example `plot_spectral_coclustering.py`.

Code for the approximation(same as above in commit):
execution time: 0.1551837921142578
```
row_diag = np.nan_to_num(row_diag)
col_diag = np.nan_to_num(col_diag)
```
Code for replacing with 1.0:
execution time: 0.1685481071472168
```
#Keeping these two as they were before the commit change
row_diag = np.where(np.isnan(row_diag), 0, row_diag)
col_diag = np.where(np.isnan(col_diag), 0, col_diag)
#Adding two new for the inf case
row_diag = np.where(np.isinf(row_diag), 1.0, row_diag)
col_diag = np.where(np.isinf(col_diag), 1.0, col_diag)
```

<img width="656" alt="Screenshot 2024-10-08 at 12 20 13" src="https://github.com/user-attachments/assets/5431e5a5-2a6b-4b67-b006-caed8f65460f">

Reply:
Note that this change is affecting both `SpectralBiclustering` and `SpectralCoclustering`. I'll have a look at the literature to understand better the impact of the normalization here.

Reply:
So the meaning of having rows or columns 0-vectors would mean that you deal with a graph where the specific vertex is isolated. In the case of a Laplacian matrix, such isolated vertex would result into a 0-eigenvalues.

Therefore, from my understanding of the problem it seems that forcing to have a 0-entry in the normalized adjacency matrix is the way to go because it will lead to a 0-eigenvalues in the subsequent SVD decomposition and this is just a sign of an isolated vertex.

Reply:
Got it, that makes sense! Would it then be useful to catch the `inf` value just as the `nan` was before and transform it into a `0` rather than `1.0`? Using the `np.nan_to_num` wouldn't be the right thing to do then.

Reply:
However, thinking about the operation below:

```python
an = row_diag[:, np.newaxis] * X * col_diag
```

It does not matter if you put 1 or 0 because the row or column in `X` is anyway 0 so it will lead to the same result I think (I have to check).

Reply:
ok noted! I will do both. thanks for looking into this so deeply. @ogrisel mentioned at the very beginning of the discussion that the module could benefit from additional tests. should that also be added or is it out of the scope of this PR?

Reply:
Actually, I would consider it in the scope: basically, we would like the case that a vertex is isolated and check that the results of the co-clustering/bi-clustering would provide the expected results. I have yet not real intuitions what it would mean in practice but I would think that we could have a toy matrix `X` having this property and check what it means in terms of results.